README.md
# Micro Data Service

Hosts vector tiles to be used in conjuction with the OSM Tasking Manager.

## Install

```bash
$ npm install
```

## Quickstart

```bash
$ npm start
```

## Create Vector tiles

Using [`tippecanoe`](https://github.com/mapbox/tippecanoe) you can generate Vector Tiles based on your GeoJSON.

**Log into server**

```bash
$ ssh root@data.osmcanada.ca
```

You will be adding the Vector Tiles `.mbtiles` inside the `/data` folder.

**Download new data**

```bash
$ cd micro-data-service/data
$ wget https://raw.githubusercontent.com/osmottawa/imports/master/CSDraveurs/CSD-schools.json
```

### Convert GeoJSON Point to Vector Tile

```bash
SOURCE=ottawa-address

# Zoom 0 to 18
tippecanoe \
    --output=$SOURCE.mbtiles \
    --force \
    --base-zoom 0 \
    --no-feature-limit \
    --no-tile-size-limit \
    --minimum-zoom 0 \
    --maximum-zoom 18 \
    --buffer 0 \
    $SOURCE.geojson
```

Restart the server.

```bash
$ docker-compose restart
```

You should now see your new dataset in the JSON.

[https://data.osmcanada.ca/datasets](https://data.osmcanada.ca/datasets)

```json
[
  "cecce-schools",
  "csd-schools",
  "oc-transpo-stops",
  "ocsb-schools",
  "ottawa-buildings"
]
```

**Tasking Manager URL**

https://data.osmcanada.ca/{z}/{x}/{y}/csd-schools.osm

**Import into JOSM URL**

[http://localhost:8111/import?new_layer=true&url=https://data.osmcanada.ca/{z}/{x}/{y}/csd-schools.osm](http://localhost:8111/import?new_layer=true&url=https://data.osmcanada.ca/{z}/{x}/{y}/csd-schools.osm)

### Polygon GeoJSON (Advanced)

```bash
SOURCE=halifax-buildings

# Zoom 13
tippecanoe \
    --output=$SOURCE-z13.mbtiles \
    --force \
    --minimum-zoom 13 \
    --maximum-zoom 13 \
    --full-detail 19 \
    --no-line-simplification \
    --no-feature-limit \
    --no-tile-size-limit \
    --no-polygon-splitting \
    --no-clipping \
    --no-duplication \
    $SOURCE.geojson

# Zoom 14
tippecanoe \
    --output=$SOURCE-z14.mbtiles \
    --force \
    --minimum-zoom 14 \
    --maximum-zoom 14 \
    --full-detail 18 \
    --no-line-simplification \
    --no-feature-limit \
    --no-tile-size-limit \
    --no-polygon-splitting \
    --no-clipping \
    --no-duplication \
    $SOURCE.geojson

# Zoom 15
tippecanoe \
    --output=$SOURCE-z15.mbtiles \
    --force \
    --minimum-zoom 15 \
    --maximum-zoom 15 \
    --full-detail 17 \
    --no-line-simplification \
    --no-feature-limit \
    --no-tile-size-limit \
    --no-polygon-splitting \
    --no-clipping \
    --no-duplication \
    $SOURCE.geojson

# Zoom 16
tippecanoe \
    --output=$SOURCE-z16.mbtiles \
    --force \
    --minimum-zoom 16 \
    --maximum-zoom 16 \
    --full-detail 16 \
    --no-line-simplification \
    --no-feature-limit \
    --no-tile-size-limit \
    --no-polygon-splitting \
    --no-clipping \
    --no-duplication \
    $SOURCE.geojson

# Zoom 17
tippecanoe \
    --output=$SOURCE-z17.mbtiles \
    --force \
    --minimum-zoom 17 \
    --maximum-zoom 17 \
    --full-detail 15 \
    --no-line-simplification \
    --no-feature-limit \
    --no-tile-size-limit \
    --no-polygon-splitting \
    --no-clipping \
    --no-duplication \
    $SOURCE.geojson

# Merge SQLite together
sqlite3 $SOURCE-z13.mbtiles '.dump' > tmp &&
    sqlite3 $SOURCE-z14.mbtiles '.dump' >> tmp &&
    sqlite3 $SOURCE-z15.mbtiles '.dump' >> tmp &&
    sqlite3 $SOURCE-z16.mbtiles '.dump' >> tmp &&
    sqlite3 $SOURCE-z17.mbtiles '.dump' >> tmp &&
    sqlite3 $SOURCE.mbtiles < 'tmp'
rm $SOURCE-z13.mbtiles $SOURCE-z14.mbtiles $SOURCE-z15.mbtiles $SOURCE-z16.mbtiles $SOURCE-z17.mbtiles tmp
```

## Configure Server

Log in to the server.

```bash
$ ssh root@data.osmcanada.ca
$ cd micro-data-service/
```

Restart Server using `docker-compose`

```bash
$ docker-compose restart
Restarting mmb ... done
```

Stop & Start the server

```bash
$ docker-compse stop
Stopping mmb ... done
```
```bash
$ docker-compose start
Starting web ... done
```

Check if service is running

```bash
$ docker ps

CONTAINER ID        IMAGE                  COMMAND             CREATED             STATUS              PORTS                    NAMES
4366f14bf554        mobilemapbuilder_web   "npm start"         19 hours ago        Up 21 seconds       0.0.0.0:5000->5000/tcp   mmb
```

Check logs

```bash
$ docker-compose logs
```


CHANGELOG.md
## 0.5.0 - 2016-11-17

- Filter results by intersecting existing OSM data based on attributes

## 0.4.2 - 2016-11-17

- Add 100 concurent HTTP connections
- Improve drasticaly Wikidata lookup (stable)
- `--radius` instead of `--distance`

## 0.3.1 - 2016-11-13

- Fix Wikidata crashing issue
- Add Wikidata automatic tagging

```
?wikidata=true
```

- Add `filter` option, uses the same syntax as [`osm-tag-stats`](https://github.com/mapbox/osm-tag-stats)

```
?filter=[["==","place","town"],["==","@user","jon_snow"]]
```

## 0.2.0 - 2016-11-11

- Switched to Vector Tiles instead of GeoJSON
- Dynamic data loading

## 0.1.0 - 2016-10-01

Project started

download.sh
#!/bin/bash

while getopts ":c:t:z:f:h" opt; do
	case $opt in
		h)
			echo "-c	Specify country" >&2
			echo >&2
			echo "-t	Specify the tag to filter. Default: place" >&2
			echo >&2
			echo "-z	Specify the maximum zoom for tippecanoe" >&2
			echo >&2
			echo "-f	Specify the output folder" >&2
			echo >&2
			echo "-h	Prints this help file" >&2
			exit 0
			;;
		\?)
			echo "Invalid option: -$OPTARG" >&2
			exit 1
			;;
		:)
			echo "Option -$OPTARG requires an argument." >&2
			exit 1
			;;
		c)
			COUNTRY=$OPTARG
			;;
		t)
			TAG=$OPTARG
			;;
		z)
			MAX_ZOOM=$OPTARG
			;;
		f)
			FOLDER=$OPTARG
			;;		
	esac
done
if [ -z ${FOLDER+x} ]; then	FOLDER="/root/micro-data-service/data"; fi
if [ -z ${TAG+x} ]; then TAG="place"; fi
if [ -z ${MAX_ZOOM+x} ]; then MAX_ZOOM=18; fi

# Download QA-Tile
# ================
if [ ! -f $COUNTRY.mbtiles ]; then
    BASE=https://s3.amazonaws.com/mapbox/osm-qa-tiles/latest
    if [ "$COUNTRY" = "planet" ]; then
        URL=$BASE.planet.mbtiles.gz
    else
        URL=$BASE.country/$COUNTRY.mbtiles.gz
    fi
    wget $URL -O $COUNTRY.mbtiles.gz
    gzip --force --decompress $COUNTRY.mbtiles.gz
fi

# OSM-Tag-Stats
# =============
FILTER="[
    \"all\",
    [\"has\", \"$TAG\"],
    [\"==\", \"@type\", \"node\"]
]"
echo $FILTER > $TAG.json
if [ ! -f $TAG-$COUNTRY.geojson ]; then
    rm -r -f tmp-osm-tag-stats/
    osm-tag-stats \
        --geojson=$TAG-$COUNTRY.geojson \
        --mbtiles=$COUNTRY.mbtiles \
        --filter=$TAG.json
fi

# Remove SQL Dumps
# ================
rm -f $TAG-$COUNTRY.dump

# Tippecanoe
# ==========
for ZOOM in $(seq 0 $MAX_ZOOM); do
    tippecanoe \
        --output=$TAG-$COUNTRY-z$ZOOM.mbtiles \
        --force \
        --minimum-zoom $ZOOM \
        --maximum-zoom $ZOOM \
        --full-detail $((16 - ZOOM + 16)) \
        --no-line-simplification \
        --no-feature-limit \
        --no-tile-size-limit \
        --no-polygon-splitting \
        --no-clipping \
        --no-duplication \
        $TAG-$COUNTRY.geojson

    # SQL Dump
    # ========
    sqlite3 $TAG-$COUNTRY-z$ZOOM.mbtiles '.dump' >> $TAG-$COUNTRY.dump
    rm $TAG-$COUNTRY-z$ZOOM.mbtiles
done

# Merge MBTiles
# =============
rm -f $TAG-$COUNTRY.mbtiles
sqlite3 $TAG-$COUNTRY.mbtiles < $TAG-$COUNTRY.dump

# Clean Files
# ===========
rm -f -r tmp-osm-tag-stats/
rm -f $TAG-$COUNTRY.dump
rm -f $COUNTRY.mbtiles.gz

# Upload to AWS
# =============
aws s3 cp $TAG-$COUNTRY.mbtiles s3://data.osmcanada.ca/$TAG-$COUNTRY.mbtiles
aws s3 cp $TAG-$COUNTRY.geojson s3://data.osmcanada.ca/$TAG-$COUNTRY.geojson

# Copy to Folder
# ==============
cp /tmp/$TAG-$COUNTRY.mbtiles $FOLDER/$TAG-$COUNTRY.mbtiles
